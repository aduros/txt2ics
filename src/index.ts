export { textToIcs } from './txt2ics'
export type { TextToIcsOptions, TextToIcsResult } from './txt2ics'
