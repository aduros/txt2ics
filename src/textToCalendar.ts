import { ICalCalendar, ICalCalendarMethod } from 'ical-generator'
import moment from 'moment'
import { hash } from 'ohash'
import { OpenAI } from 'openai'
import { zodResponseFormat } from 'openai/helpers/zod'
import { z } from 'zod'

export interface TextToCalendarOptions {
  /** The source text. */
  text: string

  /** OpenAI model. */
  model: string

  /** OpenAI client. */
  openai?: OpenAI
}

export interface TextToCalendarResult {
  calendar: ICalCalendar
}

/**
 * Converts text to iCalendar events.
 *
 * @param opts Options
 * @returns The result
 */
export async function textToCalendar(
  opts: TextToCalendarOptions,
): Promise<TextToCalendarResult> {
  // Lop off the timezone offset suffix for timestamps. GPT likes to add the timezone offset, but we
  // already handle timezones in a separate response field
  const dateTimeWithoutZone = z
    .string()
    .trim()
    .transform((dateTime) => moment.utc(dateTime.replace(/(T.*)[+-].*$/, '$1')))

  const resultSchema = z.object({
    events: z.array(
      z.object({
        title: z.string().trim().describe('The title of the event'),

        timeStart: dateTimeWithoutZone.describe(
          'The starting datetime of the event',
        ),

        timeEnd: z
          .union([dateTimeWithoutZone, z.null()])
          .describe('If provided, the ending datetime of the event'),

        timeZone: z
          .union([z.string(), z.null()])
          .describe('If provided, the timezone ID for this event'),

        allDay: z
          .boolean()
          .describe(
            'True if the event has no specific time of day, and can occur all day or at any time of day',
          ),

        recurrenceRule: z
          .union([z.string(), z.null()])
          .describe(
            'If this is a recurring event, the repeating rule string in iCalendar RRULE format',
          ),

        description: z
          .union([z.string().trim(), z.null()])
          .describe('Any extra information about the event'),

        location: z
          .union([z.string().trim(), z.null()])
          .describe('The event location'),

        emoji: z
          .union([z.string().trim(), z.null()])
          .describe('A single emoji that best describes this event'),
      }),
    ),
  })

  const openai = opts.openai ?? new OpenAI()
  const chatCompletion = await openai.beta.chat.completions.parse({
    model: opts.model,
    temperature: 0,
    seed: 0,
    messages: [
      {
        role: 'system',
        content: 'Extract calendar events from the given text',
      },
      {
        role: 'user',
        content: [
          {
            type: 'text',
            text: opts.text,
          },
        ],
      },
    ],
    response_format: zodResponseFormat(resultSchema, 'result'),
  })

  const message = chatCompletion.choices[0].message
  const result = message.parsed
  if (!result) {
    throw new Error(`Invalid response: ${message.refusal ?? 'unknown reason'}`)
  }

  if (process.env.TXT2ICS_DEBUG) {
    console.log(JSON.stringify(result, null, '  '))
  }

  const calendar = new ICalCalendar({
    prodId: 'txt2ics',
    method: ICalCalendarMethod.PUBLISH,
  })

  // The number of times we've processed each event, keyed by hash
  const eventCounts = new Map<string, number>()

  for (const event of result.events) {
    // Use the hash of this event to form its ID instead of randomizing to keep IDs stable
    // as possible between separate runs
    const eventHash = hash(event)
    const eventCount = eventCounts.get(eventHash) ?? 0
    eventCounts.set(eventHash, eventCount + 1)

    calendar.createEvent({
      id: `${eventHash}-${eventCount}`,
      start: event.timeStart,
      end: event.timeEnd,
      summary: `${event.title}${event.emoji ? ` ${event.emoji}` : ''}`,
      description:
        event.description !== event.title ? event.description : undefined,
      location: event.location,
      repeating: event.recurrenceRule,
      allDay: event.allDay,
      timezone: event.timeZone,
      floating: !event.timeZone,

      // Use a fixed TSTAMP when running in Jest to keep things testable
      stamp: process.env.NODE_ENV === 'test' ? new Date(0) : undefined,
    })
  }

  return { calendar }
}
