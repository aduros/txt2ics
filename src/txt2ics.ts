import JSON5 from 'json5'
import type { OpenAI } from 'openai'
import { z } from 'zod'
import { zodToJsonSchema } from 'zod-to-json-schema'

export interface TextToIcsOptions {
  /** The source text. */
  text: string

  /** OpenAI client. */
  openai: OpenAI

  /** OpenAI model. */
  model: string
}

export interface TextToIcsResult {
  /** The .ics iCalendar file content. */
  ics: string
}

/**
 * Converts text to iCalendar events.
 *
 * @param opts Options
 * @returns The result
 */
export async function textToIcs(
  opts: TextToIcsOptions,
): Promise<TextToIcsResult> {
  const parametersSchema = z.object({
    events: z
      .array(
        z.object({
          title: z.string().describe('The title of the event'),
          description: z
            .string()
            .optional()
            .describe('Any extra information about the event'),
          timeStart: z.string().describe('The starting datetime of the event'),
          timeEnd: z
            .string()
            .optional()
            .describe('The end datetime of the event'),
          allDay: z
            .boolean()
            .optional()
            .describe(
              'True if the event does not have a specific time and occurs all day on the dates given',
            ),
          location: z.string().optional().describe('The event location'),
          recurTimeUnit: z
            .enum([
              'minutely',
              'hourly',
              'daily',
              'weekly',
              'monthly',
              'yearly',
            ])
            .optional()
            .describe(
              'For recurring events, the unit of frequency of the recurrence',
            ),
          recurInterval: z
            .number()
            .optional()
            .describe(
              'For recurring events, the number of time units between recurrences',
            ),
          // recurWeekDays: z
          //   .array(z.enum(['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun']))
          //   .optional()
          //   .describe(
          //     'For recurring events, the days of the week the event should recur on',
          //   ),
          recurUntilTime: z
            .string()
            .optional()
            .describe(
              'For recurring events, the datetime until the event stops recurring',
            ),
        }),
      )
      .describe('Only include a single item for each event'),
  })

  const chatCompletion = await opts.openai.chat.completions.create({
    messages: [
      {
        role: 'system',
        // content: `Extract calendar events from the given text. The current datetime: ${new Date().toISOString()}`,
        content: `Extract calendar events from the given text`,
      },
      {
        role: 'user',
        content: [
          {
            type: 'text',
            text: opts.text,
          },
        ],
      },
    ],
    temperature: 0,
    tools: [
      {
        type: 'function',
        function: {
          name: 'onComplete',
          parameters: zodToJsonSchema(parametersSchema),
        },
      },
    ],
    tool_choice: {
      type: 'function',
      function: {
        name: 'onComplete',
      },
    },
    model: opts.model,
  })

  const toolCall = chatCompletion.choices[0]?.message?.tool_calls?.[0]

  if (!toolCall || toolCall.function.name !== 'onComplete') {
    throw new Error('Invalid tool call')
  }

  const result = parametersSchema.parse(
    // We use JSON5 here because sometimes OpenAI returns invalid JSON
    JSON5.parse(toolCall.function.arguments),
  )

  if (process.env.TXT2ICS_DEBUG === '1') {
    console.log(JSON.stringify(result.events, null, '  '))
  }

  const creationTime = toIcsDate(new Date().toISOString(), false)

  const ics = [
    'BEGIN:VCALENDAR',
    'PRODID:txt2ics',
    'VERSION:2.0',
    result.events
      .map((event) => {
        const lines = [
          'BEGIN:VEVENT',
          `UID:${randomId()}`,
          `SUMMARY:${oneLine(event.title)}`,
          `DTSTAMP:${creationTime}`,
          `DTSTART:${toIcsDate(event.timeStart, event.allDay)}`,
        ]
        if (event.timeEnd) {
          lines.push(`DTEND:${toIcsDate(event.timeEnd, event.allDay)}`)
        }
        if (event.location) {
          lines.push(`LOCATION:${oneLine(event.location)}`)
        }
        if (event.description) {
          lines.push(`DESCRIPTION:${oneLine(event.description)}`)
        }
        if (event.recurTimeUnit) {
          let rrule = `RRULE:FREQ=${event.recurTimeUnit.toUpperCase()}`
          if (event.recurInterval !== undefined && event.recurInterval > 1) {
            rrule += `INTERVAL=${event.recurInterval}`
          }
          if (event.recurUntilTime) {
            rrule += `UNTIL=${toIcsDate(event.recurUntilTime, event.allDay)}`
          }
          lines.push(rrule)
        }
        lines.push('END:VEVENT')
        return lines.join('\n')
      })
      .join('\n'),
    'END:VCALENDAR',
  ].join('\n')

  return {
    ics,
  }
}

function oneLine(text: string) {
  return text.split('\n')[0]
}

function randomId() {
  const base62 =
    '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'
  let str = ''
  for (let ii = 0; ii < 22; ++ii) {
    str += base62.charAt((Math.random() * 62) >>> 0)
  }
  return str
}

function pad(i: number) {
  return i < 10 ? `0${i}` : i
}

function toIcsDate(dateString: string, ignoreTime: boolean | undefined) {
  const timestamp = Date.parse(dateString)
  if (isNaN(timestamp)) {
    throw new Error(`Invalid date: ${dateString}`)
  }

  const date = new Date(timestamp)
  const year = date.getUTCFullYear()
  const month = pad(date.getUTCMonth() + 1)
  const day = pad(date.getUTCDate())

  const hour = pad(date.getUTCHours())
  const minute = pad(date.getUTCMinutes())
  const second = pad(date.getUTCSeconds())
  const timeSuffix = ignoreTime ? '' : `T${hour}${minute}${second}`
  return `${year}${month}${day}${timeSuffix}`
}
