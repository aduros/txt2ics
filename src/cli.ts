#!/usr/bin/env node --no-deprecation

import { createReadStream, createWriteStream } from 'node:fs'

import { program } from 'commander'
import OpenAI from 'openai'

import { textToIcs } from './txt2ics'

program
  .name('txt2ics')
  .description(
    'Transform plain text to an .ics file using GPT.\n\nMore info: https://github.com/aduros/txt2ics',
  )
  .option('-m, --model <model>', 'OpenAI model to use', 'gpt-4o-mini')
  .option('-o, --output <file>', 'Output file, or "-" to write to stdout', '-')
  .argument('[file]', 'Input text file, or "-" to read from stdin', '-')

  .action(
    async (
      file: string,
      opts: {
        model: string
        output: string
      },
    ) => {
      const { model } = opts
      const openai = new OpenAI()

      const inStream = file === '-' ? process.stdin : createReadStream(file)

      let text = ''
      for await (const chunk of inStream) {
        text += String(chunk)
      }

      const result = await textToIcs({
        text,
        openai,
        model,
      })

      const outStream =
        opts.output === '-' ? process.stdout : createWriteStream(opts.output)
      outStream.end(result.ics)
    },
  )

program.parse()
