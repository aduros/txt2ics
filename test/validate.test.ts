import { readFile } from 'node:fs/promises'

import { globSync } from 'glob'

import { textToCalendar } from '../src'

// Skip for now, output isn't deterministic enough to test :(
describe.skip('Validate output', () => {
  const inputFiles = globSync(`*.test.md`, { absolute: false, cwd: __dirname })

  it.each(inputFiles)(
    '%s',
    async (inputFile) => {
      const text = await readFile(`${__dirname}/${inputFile}`, 'utf8')
      const { calendar } = await textToCalendar({
        text,
        model: 'gpt-4o-2024-08-06',
      })
      expect(calendar.toString()).toMatchSnapshot()
    },
    15000,
  )
})
