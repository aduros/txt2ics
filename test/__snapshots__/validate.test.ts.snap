// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate output agenda.test.md 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-txt2ics
METHOD:PUBLISH
BEGIN:VEVENT
UID:lsxN2Tl9fO-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19690106
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=MONTHLY;BYDAY=1MO
SUMMARY:Pay rent üè†
END:VEVENT
BEGIN:VEVENT
UID:cP8oqLsvWg-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART:19690111T120000
RRULE:FREQ=WEEKLY;INTERVAL=2;BYDAY=SA
SUMMARY:Lunch with Buzz üçï
LOCATION:Luigi's Pizza
END:VEVENT
BEGIN:VEVENT
UID:SeGcZIWPDA-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19690225
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=YEARLY
SUMMARY:George's Birthday üéÇ
END:VEVENT
BEGIN:VEVENT
UID:kEZgStlXvG-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19690618
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=YEARLY
SUMMARY:Paul's Birthday üéÇ
END:VEVENT
BEGIN:VEVENT
UID:Mwa6AcpcMW-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19690707
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=YEARLY
SUMMARY:Ringo's Birthday üéÇ
END:VEVENT
BEGIN:VEVENT
UID:vtI41BBoNS-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19691009
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=YEARLY
SUMMARY:John's Birthday üéÇ
END:VEVENT
BEGIN:VEVENT
UID:6MfvOQxamH-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART:19690212T090000
DTEND:19690212T100000
SUMMARY:Dentist appointment ü¶∑
DESCRIPTION:Remember to ask doctor about braces
END:VEVENT
BEGIN:VEVENT
UID:YPPHHm2oAJ-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19690415
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
SUMMARY:Date with Uncle Sam üá∫üá∏
END:VEVENT
BEGIN:VEVENT
UID:rPe8Or8wCw-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART:19690814T200000
SUMMARY:Bus to New York üöå
END:VEVENT
BEGIN:VEVENT
UID:YOuocpMNMd-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19690815
DTEND;VALUE=DATE:19690818
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
SUMMARY:Woodstock Festival üé∏
END:VEVENT
BEGIN:VEVENT
UID:5ps7RneblV-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19691224
DTEND;VALUE=DATE:19691231
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
SUMMARY:Christmas vacation üéÑ
END:VEVENT
BEGIN:VEVENT
UID:gWYXyo3eP0-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19700101
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
SUMMARY:Quit drinking üö´üç∫
END:VEVENT
END:VCALENDAR"
`;

exports[`Validate output recurring.test.md 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-txt2ics
METHOD:PUBLISH
BEGIN:VEVENT
UID:5EMPhYPhRt-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART:19990105T150000
DTEND:19990105T170000
RRULE:FREQ=WEEKLY;BYDAY=TU,TH
SUMMARY:Scruffy's Sponge Bath üßº
DESCRIPTION:Give Scruffy a sponge bath
END:VEVENT
BEGIN:VEVENT
UID:lSoagLyTxN-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19990129
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=MONTHLY;BYDAY=-1FR
SUMMARY:Send Scruffy an Invoice üìÑ
DESCRIPTION:Send Scruffy an invoice on the last Friday of each month
END:VEVENT
BEGIN:VEVENT
UID:gW2iPzwYn8-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:19990104
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=WEEKLY;INTERVAL=2;BYDAY=MO
SUMMARY:Ask Scruffy for a Promotion üíº
DESCRIPTION:Ask Scruffy for a promotion every other Monday
END:VEVENT
END:VCALENDAR"
`;

exports[`Validate output simple.test.md 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-txt2ics
METHOD:PUBLISH
BEGIN:VEVENT
UID:Y4yYBnyftq-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:20070124
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
SUMMARY:Fix the kitchen sink üîß
END:VEVENT
BEGIN:VEVENT
UID:qXiUId8oWx-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;VALUE=DATE:20070210
DTEND;VALUE=DATE:20070212
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
SUMMARY:Buy flowers üíê
LOCATION:5th and 42nd st
DESCRIPTION:Buy flowers at that florist that hangs out on 5th and 42nd st.
END:VEVENT
END:VCALENDAR"
`;

exports[`Validate output timezones.test.md 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-txt2ics
METHOD:PUBLISH
BEGIN:VEVENT
UID:pYomY9spkk-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;TZID=America/New_York:20240704T160000
SUMMARY:Party üéâ
END:VEVENT
BEGIN:VEVENT
UID:O6R7iCOGV7-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;TZID=America/Sao_Paulo:20240907T110000
SUMMARY:Samba class üíÉ
LOCATION:Rio
END:VEVENT
BEGIN:VEVENT
UID:3oOclDUHSw-0
SEQUENCE:0
DTSTAMP:19700101T000000Z
DTSTART;TZID=Europe/London:20241226T190000
SUMMARY:Boxing Day mass ‚õ™
END:VEVENT
END:VCALENDAR"
`;
